### PR TITLE
WIP: Fix style

### PR DESCRIPTION
**Pull request by niconoe-**
Do not merge the PR right now, work in progress.

@Halleck45 @UFOMelkor This is a first step refactoring about style, code review and architecture. Aim here is to reduce the volume of the code to make it more readable and quicker by avoiding feeding RAM with useless elements. Also, some OOP architecture rework has been done to facilitate the possibility to use Dependency Injection in a close future.

Basics style and performance improvements: 
- use FQN for internals (like `\PHP_EOL` instead of `PHP_EOL`) to avoid PHP looking at the current namespace
- use real function instead of aliases (`sizeof` => `count`) for time and memory performance improvements
- Fixing lots of PHP_CodeSniffer breakings using PSR-2 standard.
- Prefer "Yoda" style for comparisons
- Add FileLevel DocBlock with the author and licence information (WIP)
- Add/Fix PHPDoc with litteral explainations for classes and methods (WIP)
- minor improvements

I'd be really honored if you can have a global review on the work that has been already done about style and refactoring in order to tell me if I can continue in this way or if you have some requirements to me.
If you miss time, you can only look at the "Configuration initialization" part, which is the most developped right now. The rest is only minor style fixes.

Thank you.

### CODE REVIEW

File: src/Hal/Metric/Consolidated.php [hunk 1]
```diff
@@ -3,6 +3,11 @@
 
 use Hal\Violation\Violation;
 
+/**
+ * Class Consolidated
```
Comment (UFOMelkor) [inline src/Hal/Metric/Consolidated.php]: Perhaps we can find something better here?

File: src/Hal/Metric/Consolidated.php [hunk 2]
```diff
@@ -82,26 +90,26 @@ public function __construct(Metrics $metrics)
...
         foreach ($avg as &$a) {
-            if (sizeof($a) > 0) {
-                $a = round(array_sum($a) / sizeof($a), 2);
+            if (\count($a) > 0) {
+                $a = \round(\array_sum($a) / \count($a), 2);
             } else {
                 $a = 0;
             }
-        }
+        } unset($a);
```
Comment (UFOMelkor) [inline src/Hal/Metric/Consolidated.php]: Should be on its own line

File: src/Hal/Metric/FileMetric.php [hunk 3]
```diff
@@ -1,7 +1,12 @@
 <?php
 namespace Hal\Metric;
 
+/**
+ * Class FileMetric
```
Comment (UFOMelkor) [inline src/Hal/Metric/FileMetric.php]: Perhaps we can finde something better here or remove this comment?

File: src/Hal/Metric/FunctionMetric.php [hunk 4]
```diff
@@ -1,7 +1,12 @@
 <?php
 namespace Hal\Metric;
 
+/**
+ * Class FunctionMetric
```
Comment (UFOMelkor) [inline src/Hal/Metric/FunctionMetric.php]: Some here

File: src/Hal/Metric/Helper/MetricClassNameGenerator.php [hunk 5]
```diff
@@ -17,7 +17,7 @@ class MetricClassNameGenerator
     public static function getName(Node $node)
     {
         return ($node instanceof Node\Stmt\Class_ && $node->isAnonymous()) ?
-            'anonymous@' . spl_object_hash($node) :
+            'anonymous@' . \spl_object_hash($node) :
```
Comment (UFOMelkor) [inline src/Hal/Metric/Helper/MetricClassNameGenerator.php]: Personally I would prefer
```php
return $condition
    ? $first
    : $second;
```

File: src/Hal/Metric/MetricException.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Hal\Metric;
+
+use LogicException;
+use Throwable;
+
+/**
+ * Class MetricException
```
Comment (UFOMelkor) [inline src/Hal/Metric/MetricException.php]: Class name information can be removed.

File: src/Hal/Metric/MetricException.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * @package Hal\Metric
+ */
+class MetricException extends LogicException
+{
+    /**
+     * Return a MetricException about a disabled Length visitor.
+     * @param int $code Integer code of the exception. Defaults to 0.
+     * @param Throwable|null $previous The previous throwable used for the exception chaining.
+     * @return $this
```
Comment (UFOMelkor) [inline src/Hal/Metric/MetricException.php]: self

File: src/Hal/Metric/MetricException.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    public static function disabledLengthVisitor($code = 0, Throwable $previous = null)
+    {
+        return new static('Please enable Legth visitor first.', $code, $previous);
+    }
+
+    /**
+     * Return a MetricException about a disabled CyclomaticComplexity visitor.
+     * @param int $code Integer code of the exception. Defaults to 0.
+     * @param Throwable|null $previous The previous throwable used for the exception chaining.
+     * @return $this
```
Comment (UFOMelkor) [inline src/Hal/Metric/MetricException.php]: self

File: src/Hal/Metric/MetricException.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    public static function disabledCyclomaticComplexityVisitor($code = 0, Throwable $previous = null)
+    {
+        return new static('Please enable CyclomaticComplexity visitor first.', $code, $previous);
+    }
+
+    /**
+     * Return a MetricException about a disabled Halstead visitor.
+     * @param int $code Integer code of the exception. Defaults to 0.
+     * @param Throwable|null $previous The previous throwable used for the exception chaining.
+     * @return $this
```
Comment (UFOMelkor) [inline src/Hal/Metric/MetricException.php]: self

File: src/Hal/Metric/MetricsVisitorTrait.php [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * (c) Jean-François Lépine <https://twitter.com/Halleck45>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Hal\Metric;
+
+/**
+ * Trait MetricsVisitorTrait
```
Comment (UFOMelkor) [inline src/Hal/Metric/MetricsVisitorTrait.php]: Can be removed

File: src/Hal/Metric/MetricsVisitorTrait.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * @package Hal\Metric
+ */
+trait MetricsVisitorTrait
+{
+    /** @var Metrics The Metrics object that will store all data analysis. */
+    protected $metrics;
+
+    /**
+     * Constructor.
```
Comment (UFOMelkor) [inline src/Hal/Metric/MetricsVisitorTrait.php]: Same here

File: src/Hal/Metric/ProjectMetric.php [hunk 12]
```diff
@@ -1,7 +1,12 @@
 <?php
 namespace Hal\Metric;
 
+/**
+ * Class ProjectMetric
```
Comment (UFOMelkor) [inline src/Hal/Metric/ProjectMetric.php]: Same here

File: src/Hal/Metric/Registry.php [hunk 13]
```diff
@@ -1,13 +1,20 @@
 <?php
 namespace Hal\Metric;
 
+/**
+ * Class Registry
```
Comment (UFOMelkor) [inline src/Hal/Metric/Registry.php]: Same here

File: src/Hal/Metric/System/Changes/GitChanges.php [hunk 14]
```diff
@@ -7,13 +7,18 @@
 use Hal\Metric\Metrics;
 use Hal\Metric\ProjectMetric;
 
+/**
+ * Class GitChanges
```
Comment (UFOMelkor) [inline src/Hal/Metric/System/Changes/GitChanges.php]: Same here

File: src/Hal/Report/Csv/Reporter.php [hunk 15]
```diff
@@ -22,8 +27,9 @@ class Reporter
 
     /**
      * Reporter constructor.
+     *
      * @param Config $config
-     * @param OutputInterface $output
+     * @param Output|OutputInterface $output
```
Comment (UFOMelkor) [inline src/Hal/Report/Csv/Reporter.php]: `Output` should be enough, shouldn't it?

File: src/Hal/Report/Html/template/relations.php [hunk 16]
```diff
@@ -19,14 +18,14 @@
             ];
         }
         $classesCp[$class['name']]['externals'][] = $ext;
-    }
+    } unset($ext);
```
Comment (UFOMelkor) [inline src/Hal/Report/Html/template/relations.php]: Should be on its own line.

**Comment by niconoe-** [comment]
Hi,

My turn to be sorry for this very late answer. :disappointed: 

Since I have no more time on this project due to some personal changes, I'm not sure I'll be able to continue this fixing-style PR.

But you can try to do it.

About the points you said:

* The [false, true][condition] syntax is a bit hard to read on first look, but it offers ability to do an action without any condition, and so it reduce the algorithm complexity (by reducing also the readability, I agree). If you're not familiar with this, you can switch back to if/else or ternary conditions if you like.

* The fluent style might be a bad practice when excessive, but it might be good for readability too, in certain conditions... If your code call is always like "$a->do1(); $a->do2(); $a->do3();", you should consider making the class instanciated into $a having fluent methods. Otherwise, you can keep returning void in order to allow you returning something else in a feature improvement without having a BC break.

* You could definitly remove the uiseless value comments. I used automated comments sometimes and it creates stupid comments.

I could try to work back again on this project but as I didn't follow the evolution of the last 7 months, I should either restart all from the 2.4.0 version, or waiting for you to merge correctly (without taking care of the style) in my branch so I can continue it.
If you do not want to resolve the conflict, I'll restart again, but I will probably fix each style issue separatly, to have a smaller PR :smile:.

Thank you for your big review.
Regards,

**Comment by gmponos** [comment]
Hello there and thanks for the contribution.

>  I will probably fix each style issue separatly, to have a smaller PR 😄.

Let me know if there is still any interest contributing with smaller PR. My intention is to add a php-cs-fixer in order to automatically handle some style changes.

**Comment by gmponos** [comment]
Because of all the conflicts I will close this.. @niconoe- as you said this might need reboot and split into smaller PRs.

**Pull request closed without merging**